File: solutions/code-carl/ch09_dp/T3_LC0746_MinCostClimbingStairs.java
/**
 * Practice template for:
 * - LeetCode 746. Min Cost Climbing Stairs
 *   https://leetcode.cn/problems/min-cost-climbing-stairs/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/20 15:24
 * @comment
 */
public class T3_LC0746_MinCostClimbingStairs {
    public int minCostClimbingStairs(int[] cost) {
        int prepre = 0, pre = 0, curr = 0;
        for (int i = 2; i <= cost.length; i++) {
            curr = Math.min(prepre + cost[i - 2], pre + cost[i - 1]);
            prepre = pre;
            pre = curr;
        }
        return pre;
    }
}

File: solutions/python/code-carl/ch09_dp/t3_lc0746_min_cost_climbing_stairs.py
"""LeetCode 746. Min Cost Climbing Stairs
https://leetcode.cn/problems/min-cost-climbing-stairs/
"""

from typing import List


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        if len(cost) == 0:
            return 0
        elif len(cost) == 1:
            return cost[0]
        elif len(cost) == 2:
            return min(cost[0], cost[1])

        dp: List[int] = [0] * (len(cost) + 1)
        for i in range(2, len(dp)):
            dp[i] = min(dp[i - 1] + cost[i - 1], dp[i - 2] + cost[i - 2])

        return dp[-1]


if __name__ == "__main__":
    solution = Solution()
    assert solution.minCostClimbingStairs([10, 15, 20]) == 15
    assert solution.minCostClimbingStairs([1, 100, 1, 1, 1, 100, 1, 1, 100, 1]) == 6
    print("All tests passed.")
